openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py

```diff
@@ -0,0 +1,156 @@
+            nargs=2,
+            help=('The edX OAuth credentials in the command line.  The first '
+                  'argument is the OAuth secret, the second argument is the '
+                  'OAuth key. Notice that command line arguments are insecure, '
+                  'see `--oauth-creds-file`.'),
+        )
+        subparser.add_parser(
+            'modulestore',
+            help=("Use a local modulestore intsance to retrieve blocks database on "
+                  "the instance where the command is being run.  You don't need "
+                  "to enable API access.")
+        )
+
+    def write_error(self, *args, **kwds):
+        """
+        Write error messages to stdout.
+        """
+        return self.stdout.write(self.style.ERROR(*args, **kwds))
+
+    def write_success(self, *args, **kwds):
+        """
+        Write success messages to stdout.
+        """
+        return self.stdout.write(self.style.SUCCESS(*args, **kwds))
+
+    def handle(self, *args, **options):
+        """
+        Collect all blocks from a course that are "importable" and write them to the
+        a blockstore library.
+        """
+
+        # Search for the library.
+
+        try:
+            contentlib_api.get_library(options['library-key'])
+        except contentlib_api.ContentLibraryNotFound as exc:
+            raise CommandError("The library specified does not exist: "
+                               f"{options['library-key']}") from exc
+
+        # Validate the method and its arguments, instantiate the openedx client.
+
+        if options['method'] == 'api':
+            if options['oauth_creds_file']:
+                with options['oauth_creds_file'] as creds_f:
+                    oauth_key, oauth_secret = [v.strip() for v in creds_f.readlines()]
+            elif options['oauth_creds']:
+                oauth_key, oauth_secret = options['oauth_creds']
+            else:
+                raise CommandError("Method 'api' requires one of the "
+                                   "--oauth-* options, and none was specified.")
+            edx_client = contentlib_api.EdxApiImportClient(
+                options['lms_url'],
+                options['studio_url'],
+                oauth_key,
+                oauth_secret,
+                library_key=options['library-key'],
+            )
+        elif options['method'] == 'modulestore':
+            edx_client = contentlib_api.EdxModulestoreImportClient(
+                library_key=options['library-key'],
+            )
+        else:
+            raise CommandError(f"Method not supported: {options['method']}")
+
+        failed_blocks = []
+
+        def on_progress(block_key, block_num, block_count, exception=None):
+            self.stdout.write(f"{block_num}/{block_count}: {block_key}: ", ending='')
+            if exception:
+                self.write_error('❌')
+                self.stderr.write(f"Failed to import modulestore block: {exception}")
```

Comment:
Is that a common convention? (and if so, why is the functionality not already baked into the superclass here?) If so, then we're all good. If not super common, please just include a comment explaining it for the dummies like me :)

Reply:
> Is that a common convention?

Talking about the *nix command line in general? I've seen it many times, although I am not entitled to say, "this is a common convention around all command lines in the world". I guess it would be a point for style and discussion. I've seen many commands regard `stderr` as a place to log errors and diagnostic-related information, while `stdout` is used to show expected output or summarized progress. One usage pattern is: You call the command with `foobar 2>errors.log`, get a cleaner output, easier to see the big picture, and if something went wrong and you want to dive into the details of the errors, you check stderr. It is advantageous when your command will be used in scripts or automated environments. BTW, this is what I mean when I said "`2>/dev/null`-style calls", although redirecting to `/dev/null`, you are pretty much ignoring the details. My point was, you can redirect stderr to another place other than stdout.

> why is the functionality not already baked into the superclass here?

It is. They allow you to write to different output streams and allows you to pretty-print the output as well. I just added two wrappers called `write_error(...)` and `write_success(...)` to colorize the output according to upstream's convention of what would be a good color for error and success. I could just have used `self.stdout.write(self.style.ERROR(...))` instead. I think it is a little bit longer and repetitive, hence the wrapper method.

> If not super common, please just include a comment explaining it

I'll get rid of the helper method, which will make it explicit and less confusing.